### DISCUSSION TITLE
[Bug]: Misleading text about shared references

### DISCUSSION THREAD

**Discussion by last-partizan** [olimorris/codecompanion.nvim#464] [Ideas]
### Your `minimal.lua` config

irrelevant

### Log output

irrelevant

### Describe the bug

It's not really a breaking bug, but more like how the plugin displays shared variables.

`> Sharing: ...` is added to the next message, but in it does not actually adds any variables to the second message.

Expected behavior is probably - display variables only on the original message.

<pre>
## Me ─

> Sharing:
> - /tmp/test.txt

#buffer

I'm testing how sharing context works. Display me received file. Short answer, no suggestions on next steps.

## copilot ─

```text
Version 1
```

## Me ─

> Sharing:
> - /tmp/test.txt

How about now?

## copilot ─

```text
Version 1
```
</pre>

Json log from `gd`:

```lua
--Settings
adapter = "copilot"
n = 1
model = "claude-3.5-sonnet"
max_tokens = 4096
temperature = 0
top_p = 1

--Messages
{ {
    content = "You are an AI programming assistant named \"CodeCompanion\".\nYou are currently plugged in to the Neovim text editor on a user's machine.\n\nYour core tasks include:\n- Answering general programming questions.\n- Explaining how the code in a Neovim buffer works.\n- Reviewing the selected code in a Neovim buffer.\n- Generating unit tests for the selected code.\n- Proposing fixes for problems in the selected code.\n- Scaffolding code for a new workspace.\n- Finding relevant code to the user's query.\n- Proposing fixes for test failures.\n- Answering questions about Neovim.\n- Running tools.\n\nYou must:\n- Follow the user's requirements carefully and to the letter.\n- Keep your answers short and impersonal, especially if the user responds with context outside of your tasks.\n- Minimize other prose.\n- Use Markdown formatting in your answers.\n- Include the programming language name at the start of the Markdown code blocks.\n- Avoid line numbers in code blocks.\n- Avoid wrapping the whole response in triple backticks.\n- Only return code that's relevant to the task at hand. You may not need to return all of the code that the user has shared.\n- Use actual line breaks instead of '\\n' in your response to begin new lines.\n- Use '\\n' only when you want a literal backslash followed by a character 'n'.\n- All non-code responses must use English.\n\nWhen given a task:\n1. Think step-by-step and describe your plan for what to build in pseudocode, written out in great detail, unless asked not to do so.\n2. Output the code in a single code block, being careful to only return relevant code.\n3. You should always generate short suggestions for the next user turns that are relevant to the conversation.\n4. You can only give one reply for each conversation turn.",
    id = 1947269257,
    opts = {
      visible = false
    },
    role = "system"
  }, {
    content = "I'm testing how sharing context works. Display me received file. Short answer, no suggestions on next steps.",
    id = -546777357,
    opts = {
      visible = true
    },
    role = "user"
  }, {
    content = "Buffer ID: 1\nName: test.txt\nPath: /tmp/test.txt\nFiletype: text\nContent:\n```text\n1:  Version 1\n```\n",
    id = -454189760,
    opts = {
      reference = "/tmp/test.txt",
      tag = "variable",
      visible = false
    },
    role = "user"
  }, {
    content = "```text\nVersion 1\n```",
    id = 1226543241,
    opts = {
      visible = true
    },
    role = "llm"
  }, {
    content = "How about now?",
    id = 419434627,
    opts = {
      visible = true
    },
    role = "user"
  }, {
    content = "```text\nVersion 1\n```",
    id = 1226543241,
    opts = {
      visible = true
    },
    role = "llm"
  } }
```

### Reproduce the bug

1. Share a `#buffer` with agent
2. Look as "> Sharing" in second message (unexpected)

### Final checks

- [X] I have made sure this issue exists in the latest version of the plugin
- [X] I have tested with the `minimal.lua` file from above and have shared this
- [X] I have shared the contents of the log file

**Comment by GitMurf** [comment]
I love the fact that it does not re-share all the buffer content over and over each message (to keep the context size "leaner"). But I agree there needs to be a way to "update" the content when you want to after changing the buffer. What if a timestamp gets added to the reference for the buffer so you can see when it was shared (and thus the time of content being used from that buffer) and if you include again in a message later the `#buffer` tag then that means you would like to update the content so it would add an updated copy of the buffer and then would update the timestamp of the buffer in the references list.

thoughts?

**Comment by GitMurf** [comment]
So maybe the references say something like:

> Sharing:
> - /tmp/test.txt (as of 10:15 AM)

**Comment by GitMurf** [comment]
In a perfect world what would be VERY cool is if using the same timestamp concept, the references could highlight / flag any shared reference items that have recent changes... meaning if you update a buffer file and save, the references simply check the last modified timestamp of the reference files and compare to its timestamp when shared and can let you know if a reference has changed to prompt you to re-include an update if you want to.

**Comment by olimorris** [comment]
> \> Sharing: ... is added to the next message, but in it does not actually adds any variables to the second message.

My logic for this is that once you add a reference (i.e. a buffer or a file) to the chat buffer, it becomes part of the payload which is sent with _every_ request to your LLM. I chose the verb "sharing" because that's exactly what's happening with every message you send. If I'd have used ` > Added ...` then I think that could be interpreted to mean the reference was added with that message.

There's a fine line to tread in the chat buffer between useful information and visual bloat. This was a feature I hadn't intended to add but acknowledged that for some users it was really useful.

I'm also planning a lot of features related to this e.g. sticky/updateable references on every request and the ability to modify references on the fly.

Closing this as it's not a bug but will transfer to a discussion.

**Reply by GitMurf** [reply]
> I'm also planning a lot of features related to this e.g. sticky/updateable references on every request and the ability to modify references on the fly.

With this along with the plan for ability to remove references, I think that gives us all the tools we need and should resolve the concern from the original gh issue. Looking forward to it @olimorris !

**Reply by last-partizan** [reply]
> My logic for this is that once you add a reference (i.e. a buffer or a file) to the chat buffer, it becomes part of the payload which is sent with every request to your LLM. I chose the verb "sharing" because that's exactly what's happening with every message you send. If I'd have used  > Added ... then I think that could be interpreted to mean the reference was added with that message.

I think, I understand ...

By "sent on every request", you mean that we send full thread to the LLM. E.q.

```
- User: here's my file
- Agent: ...
- User: ...
```

And we send this entire dialog to the LLM (including a file at the beginning).

I agree with @GitMurf on this:

> I love the fact that it does not re-share all the buffer content over and over each message (to keep the context size "leaner"). 

But, original problem is: on each subsequent User message added a quote about "Sharing ...". This is not neccesary, as it gives the wrong impression that buffer is actually updated on each message.

For me - it's great that it works as it works now, the only minor annoyance is this "Sharing" in every message.
On top of that, if i use `#buffer` command again, it become:


```
    Sharing:
        /tmp/test.txt
        /tmp/test.txt
```

Looking at the commit: https://github.com/olimorris/codecompanion.nvim/commit/767efee2ff8fd7d84ac9a5ff35bac52eecec5584

```diff
@@ -801,7 +803,7 @@ function Chat:submit(opts)
  if not opts.regenerate and not vim.tbl_isempty(message) then
    self:add_message({ role = config.constants.USER_ROLE, content = message.content })
  end
-  message = self.messages[#self.messages]
+  message = self.References:clear(self.messages[#self.messages])
```

It seems like the references should be cleared after submit. But somehow they're cleared only from "sending to LLM", but not from "Displaying to a user".

Maybe this a bug?

**Reply by olimorris** [reply]
> By "sent on every request", you mean that we send full thread to the LLM. E.q.
> 
> ```
> - User: here's my file
> - Agent: ...
> - User: ...
> ```

Yes. This is true for every LLM which the plugin supports. You send the _full_ history alongside every new message. So you're always "sharing" the original `#buffer` despite it being n - x messages ago.

> But, original problem is: on each subsequent User message added a quote about "Sharing ...". This is not neccesary, as it gives the wrong impression that buffer is actually updated on each message.
> It seems like the references should be cleared after submit. But somehow they're cleared only from "sending to LLM", but not from "Displaying to a user".

You're still "sharing" that buffer in subsequent messages. Therefore the wording `sharing: > - init.lua` is factually correct for all new messages. I understand how you may have the impression of it updating on each message but I can clarify, this is not the case...yet.

> Maybe this a bug?

No, this is the intended design as you can see in the code.

You can turn off the displaying of references in the chat buffer with `display.chat.show_references = false` in your config.

**Reply by last-partizan** [reply]
> No, this is the intended design as you can see in the code.

But, why does it needs to be displayed on every subsequent message?

In UI each `-- Me --` block is supposed to represent a single user message to the LLM.
It makes sense that the message with `#buffer` will include this text about `Sharing ...`. But I don't understand what's point of displaying it for next message?

To be crystal clear, it makes sense to display `> Sharing ..` when rendering this message:

```lua
{
    content = "Buffer ID: 1\nName: test.txt\nPath: /tmp/test.txt\nFiletype: text\nContent:\n```text\n1:  Version 1\n```\n",
    id = -454189760,
    opts = {
      reference = "/tmp/test.txt",
      tag = "variable",
      visible = false
    },
    role = "user"
  }
```

But for this, it's unexpected.

```lua
{
    content = "How about now?",
    id = 419434627,
    opts = {
      visible = true
    },
    role = "user"
  }
```

**Reply by last-partizan** [reply]
Oh, wait. I totally missed this.

> You're still "sharing" that buffer in subsequent messages. 

How does it work?

**Reply by olimorris** [reply]
Press `gd` in the chat buffer. You'll see the payload that gets sent to the LLM with _every_ request. Your buffer is in that. Therefore it's shared with every response.

I toggle chat buffers quite regularly. Coming back to a chat buffer after sometime and seeing what context I've shared with the LLM is useful to me. The same with vim help docs. The same with URLs. It's useful to know the LLM is citing those resources in its responses.

To be blunt, I won't be changing this functionality or how references are displayed in the chat buffer.

**Reply by GitMurf** [reply]
> But, why does it needs to be displayed on every subsequent message?

Because how else do you know what is in the conversation history? Without it you won't know / remember what is part of the context as the entire conversation is always the context. If I can't remember whether I already shared file1.txt and file2.txt I will not know the context AI has overall which is important.

**Reply by last-partizan** [reply]
Yeah, it's probably matter of preference.

![image](https://github.com/user-attachments/assets/c78856a6-019f-45e1-bb17-29f987bae790)

I just like how Zed does it. It renders a file shared only for original message.

> To be blunt, I won't be changing this functionality or how references are displayed in the chat buffer.

Would you accept PR adding an option to display references only for the original message?
